src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,713 @@
+					mapi++;
+				char name[32];
+				input_item_id id = ITEM_ID_INVALID;
+				name[0] = 0;
+				if(key_map[mapi].evdev_id == i) {
+					strncpy(name, key_map[mapi].name, 31);
+					name[31] = 0;
+					if((id = key_map[mapi].mame_id) != ITEM_ID_INVALID)
+						numkeys++;
+				} else {
+					if(i<BTN_MISC)
+						snprintf(name, 32, "KEYCODE %d", i);
+					else
+						snprintf(name, 32, "BUTTON %d", numbuttons+1);
+				}
+				if(id == ITEM_ID_INVALID && numbuttons<32)
+					id = static_cast<input_item_id>(ITEM_ID_BUTTON1 + numbuttons++);
+				if(id != ITEM_ID_INVALID) {
+					device()->add_item(name, id, generic_button_get_state<uint8_t>, keystate+i);
+				}
+			}
+
+		if(fd < 0) {
+			char fname[16];
+			snprintf(fname, 16, "event%d", info.num);
+			fd = openat(dirfd, fname, O_RDONLY|O_NONBLOCK);
+			if(fd < 0) {
+				osd_printf_verbose("evdev: unable to open %s: %s\n", fname, strerror(errno));
+			}
+		}
+
+		mapi = 0;
+		for(int i=0; i<REL_MAX; i++)
+			if(info.relaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || relaxis_map[mapi].evdev_id) && relaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(relaxis_map[mapi].evdev_id == i) {
+					device()->add_item(relaxis_map[mapi].name, relaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, relaxis+i);
+				}
+			}
+
+		mapi = 0;
+		for(int i=0; i<ABS_MAX; i++)
+			if(info.absaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || absaxis_map[mapi].evdev_id) && absaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(absaxis_map[mapi].evdev_id == i) {
+					if(fd >= 0)
+						ioctl(fd, EVIOCGABS(i), &absaxis[i].absinfo);
+					device()->add_item(absaxis_map[mapi].name, absaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, &absaxis[i].normalized);
+				}
+			}
+	}
+
+	void poll(void) override
+	{
+		static input_event ie[256]; // try to consume as many as possible at once
+		while(fd >= 0) {
+			ssize_t rd = read(fd, ie, 256*sizeof(input_event));
+			if(rd < 0) {
+				if(errno==EWOULDBLOCK || errno==EAGAIN)
+					return;
+				// cope with errors here
+			}
+			if(rd < sizeof(input_event))
+				return;
+			int nie = rd/sizeof(input_event);
+			for(int i=0; i<nie; i++) {
+				switch(ie[i].type) {
+					case EV_KEY:
+						keystate[ie[i].code] = (ie[i].value>0)? 0x80: 0;
+						break;
+					case EV_REL:
+						relaxis[ie[i].code] += ie[i].value;
+						break;
+					case EV_ABS: {
+						  auto& axis = absaxis[ie[i].code];
+						  axis.normalized = normalize_absolute_axis(ie[i].value, axis.absinfo.minimum, axis.absinfo.maximum);
+						}
+						break;
+				}
+			}
+		}
+	}
+
+	void reset(void) override
+	{
+		memset(keystate, 0, KEY_MAX);
+		memset(relaxis, 0, REL_MAX*sizeof(int32_t));
+		memset(absaxis, 0, ABS_MAX*sizeof(absaxis[0]));
+	}
+};
+
+#define MAX_EVDEV_DEVICES 64
+
+class evdev_input
+{
+private:
+	bool initialized;
+
+public:
+	int  devinput;
+	evdev_device_info dev[MAX_EVDEV_DEVICES]; // 640k should be enough for anyone
```

Comment:
Why not use s vector, or if the items can’t be moved, a vector of smart pointers or a list?

Reply:
You can never really scrub away the old dinosaur C programmer bad habits away.

Yeah, this absolutely should be a vector which also does away neatly with the silly arbitrary limit.